06-kotlin-mqtt/src/test/kotlin/KotlinSocketTest.kt

```diff
@@ -0,0 +1,98 @@
+            .from.toml.resource("config.toml")
+    private val protelisModuleName = config[ProtelisConfigSpec.protelisModuleName]
+    private val iterations = config[ProtelisConfigSpec.iterations]
+    private val nodes = config[ProtelisConfigSpec.nodes]
+    private val leaders = config[ProtelisConfigSpec.nodes]
+            .filter { it.leader }
+            .map { it.id }
+
+    override fun beforeSpec(spec: Spec) {
+        initServer()
+        nodes.forEach {
+            val mqttNetworkManager = MqttNetworkManager(IntDeviceUID(it.id), it.neighbors).apply { listen(it.listen) }
+            val program = ProtelisLoader.parse(protelisModuleName)
+            val s = spyk(ConsoleSpeaker())
+            val d = Device(program, it.id, mqttNetworkManager, s)
+            if (it.leader) {
+                d.deviceCapabilities.executionEnvironment.put("leader", true)
+            }
+            devices += d
+            speakers += s
+        }
+        repeat(iterations) {
+            devices.forEach { it.runCycle() }
+        }
+        devices.forEach { (it.netmgr as MqttNetworkManager).stop() }
+    }
+
+    override fun afterSpec(spec: Spec) {
+        super.afterSpec(spec)
+        closeServer()
+    }
+
+    fun initServer() {
+        server = Server().apply { startServer() }
+    }
+
+    fun closeServer() {
+        server.stopServer()
+    }
+
+    init {
+
+        "There should be at least 1 leader" {
+            leaders.size shouldBeGreaterThan 0
+        }
+
+        "The leader count should be correct" {
+            val messages = generateSequence(3f) { it - 1 }
+                    .take(3)
+                    .map { "The leader's count is: $it" }
+                    .toList()
+            leaders.stream()
+                    .flatMap({ x -> messages.stream().map({ msg -> Pair<Int, String>(x, msg) }) })
+                    .forEach { id ->
+                        verify(exactly = 1) { speakers[id.first].announce((id.second)) }
+                    }
+            val leftovers = config[ProtelisConfigSpec.iterations] - 3
+            if (leftovers > 0) {
+                leaders.forEach { id ->
+                    verify(exactly = leftovers) { speakers[id].announce("The leader's count is: 0.0") }
+                }
+            }
+        }
+
+        "The leaders should announce their id" {
+            leaders.forEach {
+                id -> verify(exactly = iterations) {
+                    speakers[id].announce("The leader is at $id")
+                }
+            }
+        }
+
+        "The leader neighbors should say something" {
+            leaders
+                    .flatMap { listOf(
+                            (it + nodes.size - 1) % nodes.size,
+                            (it + 1) % nodes.size) }
+                    .distinct()
+                    .forEach { id -> verify(atLeast = 1) { speakers[id].announce("Hello from the leader to its neighbor at $id") } }
+        }
+    }
+}
```

Comment:
There is no Kdoc in this file. Can you provide it for the public members?

Reply:
What should I document in a test class?

Reply:
What the tests test. Test code is as important as production code.